### PR TITLE
[Networking] Sets GossipSub RPC Inspector mandatory default

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This PR improves how the libp2p node, a component involved in network communication, is structured. Before the update, the RPC inspector suite, a vital tool for logging and monitoring network communication, was an optional add-on to the libp2p node. Now, it's a mandatory part of the libp2p node.

What does this mean? The RPC inspector suite helps keep an eye on the network communications, ensuring they're healthy and secure. It's particularly useful for logging data and investigating suspicious activities. With the changes made, instead of setting a specific RPC inspector suite, you can replace its underlying mechanism (a factory). If you don’t provide one, don't worry – the system will use a default one.

Another key change is organization and simplification. Before, the libp2p node and its components were intertwined complicatedly. The RPC inspector suite was closely connected to many other parts, which made the system harder to work with. This update neatly packs the RPC inspector suite inside the gossipsub builder. This, in turn, is neatly packed inside the libp2p node. This new arrangement makes the code cleaner and easier to understand and maintain.

Additionally, this update lays the foundation that the libp2p node itself listens for important protocol-level events (i.e., cluster change) and passes the information to the necessary components. This makes things more efficient and takes some burden off the Flow node builder.

### CODE REVIEW

File: network/p2p/p2pbuilder/gossipsub/gossipSubBuilder.go [hunk 1]
```diff
@@ -194,25 +281,31 @@ func (g *Builder) Build(ctx irrecoverable.SignalerContext) (p2p.PubSubAdapter, p
...
-		gossipSubConfigs.WithInspectorSuite(g.rpcInspectorSuite)
+	inspectorSuite, err := g.rpcInspectorSuiteFactory(
+		g.logger,
+		g.sporkId,
+		g.rpcInspectorConfig,
+		g.metricsCfg.Metrics,
+		g.metricsCfg.HeroCacheFactory,
+		g.networkType,
+		g.idProvider)
+	gossipSubConfigs.WithInspectorSuite(inspectorSuite)
```
Comment (peterargue) [inline network/p2p/p2pbuilder/gossipsub/gossipSubBuilder.go:292]: should this be called after the error check?

**Review by peterargue — APPROVED** [review]
thanks for simplifying this!

File: network/p2p/p2pbuilder/gossipsub/gossipSubBuilder.go [hunk 2]
```diff
@@ -11,10 +11,20 @@ import (
...
+	"github.com/onflow/flow-go/module/mempool/queue"
+	"github.com/onflow/flow-go/module/metrics"
+	"github.com/onflow/flow-go/network"
 	"github.com/onflow/flow-go/network/channels"
 	"github.com/onflow/flow-go/network/p2p"
+	"github.com/onflow/flow-go/network/p2p/distributor"
+	"github.com/onflow/flow-go/network/p2p/inspector"
+	"github.com/onflow/flow-go/network/p2p/inspector/validation"
+	p2pconfig "github.com/onflow/flow-go/network/p2p/p2pbuilder/config"
+	inspector2 "github.com/onflow/flow-go/network/p2p/p2pbuilder/inspector"
```
Comment (kc1116) [inline network/p2p/p2pbuilder/gossipsub/gossipSubBuilder.go:26]:
```suggestion
	inspectorbuilder "github.com/onflow/flow-go/network/p2p/p2pbuilder/inspector"
```

File: network/p2p/p2pbuilder/gossipsub/gossipSubBuilder.go [hunk 3]
```diff
@@ -131,44 +132,130 @@ func (g *Builder) SetRoutingSystem(routingSystem routing.Routing) {
...
+			inspectMsgQueueCacheCollector,
+			clusterPrefixedCacheCollector,
+			idProvider,
+			gossipSubMetrics,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create new control message valiadation inspector: %w", err)
+		}
+
+		return inspector2.NewGossipSubInspectorSuite([]p2p.GossipSubRPCInspector{metricsInspector, rpcValidationInspector}, notificationDistributor), nil
```
Comment (kc1116) [inline network/p2p/p2pbuilder/gossipsub/gossipSubBuilder.go:255]:
```suggestion
		return inspectorbuilder.NewGossipSubInspectorSuite([]p2p.GossipSubRPCInspector{metricsInspector, rpcValidationInspector}, notificationDistributor), nil
```

**Review by kc1116 — APPROVED** [review]

File: network/p2p/p2pbuilder/gossipsub/gossipSubBuilder.go (hunk 1 above)
Comment (yhassanzadeh13) [reply network/p2p/p2pbuilder/gossipsub/gossipSubBuilder.go]: https://github.com/onflow/flow-go/pull/4511/commits/38ed00e2f89e21065bfecb4996d3fadb9384ed63

File: config/config.go [hunk 4]
```diff
@@ -184,12 +184,45 @@ func LogConfig(logger *zerolog.Event, flags *pflag.FlagSet) map[string]struct{}
...
+// full key in the viper config store. This is required because in our config.yml file all configuration values for the
+// Flow network are stored one level down on the network-config property. When the default config is bootstrapped viper will
+// store these values with the "network-config." prefix on the config key, because we do not want to use CLI flags like --network-config.networking-connection-pruning
+// to override default values we instead use cleans flags like --networking-connection-pruning and create an alias from networking-connection-pruning -> network-config.networking-connection-pruning
+// to ensure overrides happen as expected.
+// Args:
+// *viper.Viper: instance of the viper store to register network config aliases on.
+// Returns:
+// error: if a flag does not have a corresponding key in the viper store.
+func SetAliases(conf *viper.Viper) error {
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply config/config.go:203]: It causes a circular dependency then between `conf` and `netconf` packages.

File: network/p2p/p2pbuilder/gossipsub/gossipSubBuilder.go (hunk 2 above)
Comment (yhassanzadeh13) [reply network/p2p/p2pbuilder/gossipsub/gossipSubBuilder.go]: https://github.com/onflow/flow-go/pull/4511/commits/4eff362ee754bf332e9ca0b621e47657c52a7ee8

File: network/p2p/p2pbuilder/gossipsub/gossipSubBuilder.go (hunk 3 above)
Comment (yhassanzadeh13) [reply network/p2p/p2pbuilder/gossipsub/gossipSubBuilder.go]: https://github.com/onflow/flow-go/pull/4511/commits/4eff362ee754bf332e9ca0b621e47657c52a7ee8

**Comment by yhassanzadeh13** [comment]
bors merge

**Comment by yhassanzadeh13** [comment]
bors merge

**Pull request merged**